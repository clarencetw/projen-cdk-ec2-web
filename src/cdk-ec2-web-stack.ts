import * as path from 'path';
import * as autoscaling from '@aws-cdk/aws-autoscaling';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as assets from '@aws-cdk/aws-s3-assets';
import * as cdk from '@aws-cdk/core';

export class CdkEc2WebStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const vpc = ec2.Vpc.fromLookup(this, 'VPC', {
      isDefault: true,
    });

    const host = new ec2.BastionHostLinux(this, 'BastionHost', {
      vpc,
      subnetSelection: { subnetType: ec2.SubnetType.PUBLIC },
    });
    host.allowSshAccessFrom(ec2.Peer.ipv4('114.114.192.168/32'));

    const asset = new assets.Asset(this, 'Asset', { path: path.join(__dirname, '../ec2-configure/configure.sh') });
    const asg = new autoscaling.AutoScalingGroup(this, 'ASG', {
      vpc,
      minCapacity: 1,
      maxCapacity: 5,
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T3,
        ec2.InstanceSize.NANO,
      ),
      machineImage: new ec2.AmazonLinuxImage({
        generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
      }),
    });
    asg.connections.allowFrom(host, ec2.Port.tcp(22));

    const localPath = asg.userData.addS3DownloadCommand({
      bucket: asset.bucket,
      bucketKey: asset.s3ObjectKey,
    });
    asg.userData.addExecuteFileCommand({
      filePath: localPath,
      arguments: '--verbose -y',
    });
    asset.grantRead(asg.role);

    const lb = new elbv2.ApplicationLoadBalancer(this, 'LB', {
      vpc,
      internetFacing: true,
    });
    const listener = lb.addListener('Listener', {
      port: 80,
    });
    listener.addTargets('Targets', {
      port: 80,
      targets: [asg],
    });

    asg.scaleOnCpuUtilization('CpuUtilization', {
      targetUtilizationPercent: 50,
    });
    asg.scaleOnIncomingBytes('IncomingBytes', {
      targetBytesPerSecond: 10 * 1024 * 1024,
    });
    asg.scaleOnOutgoingBytes('OutgoingBytes', {
      targetBytesPerSecond: 10 * 1024 * 1024,
    });
    asg.scaleOnRequestCount('RPS', {
      targetRequestsPerSecond: 1000,
    });

    new cdk.CfnOutput(this, 'PHPInfo', {
      value: `http://${lb.loadBalancerDnsName}/phpinfo.php`,
    });
  }
}
