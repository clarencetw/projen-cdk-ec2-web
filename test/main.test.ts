import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { CdkEc2WebStack } from '../src/cdk-ec2-web-stack';

test('Snapshot', () => {
  const app = new App();
  const env = { region: 'us-east-1', account: '123456789012' };
  const stack = new CdkEc2WebStack(app, 'test', { env });

  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});