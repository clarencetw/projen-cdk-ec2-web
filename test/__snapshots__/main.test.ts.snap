// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "BastionHostBastionHostIdC743CBD6": Object {
      "Description": "Instance ID of the bastion host. Use this to connect via SSM Session Manager",
      "Value": Object {
        "Ref": "BastionHost30F9ED05",
      },
    },
    "PHPInfo": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "http://",
            Object {
              "Fn::GetAtt": Array [
                "LB8A12904C",
                "DNSName",
              ],
            },
            "/phpinfo.php",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeArtifactHash1A7F30FD": Object {
      "Description": "Artifact hash for asset \\"b555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770cee\\"",
      "Type": "String",
    },
    "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3Bucket478064AC": Object {
      "Description": "S3 bucket for asset \\"b555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770cee\\"",
      "Type": "String",
    },
    "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9": Object {
      "Description": "S3 key for asset version \\"b555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770cee\\"",
      "Type": "String",
    },
    "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-gp2",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": Object {
    "ASG46ED3070": Object {
      "Properties": Object {
        "LaunchConfigurationName": Object {
          "Ref": "ASGLaunchConfigC00AF12B",
        },
        "MaxSize": "5",
        "MinSize": "1",
        "Tags": Array [
          Object {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "test/ASG",
          },
        ],
        "TargetGroupARNs": Array [
          Object {
            "Ref": "LBListenerTargetsGroup76EF81E8",
          },
        ],
        "VPCZoneIdentifier": Array [
          "p-12345",
          "p-67890",
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": Object {
        "AutoScalingScheduledAction": Object {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
    },
    "ASGInstanceProfile0A2834D7": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "ASGInstanceRoleE263A41B",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "ASGInstanceRoleDefaultPolicy7636D8BF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3Bucket478064AC",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3Bucket478064AC",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ASGInstanceRoleDefaultPolicy7636D8BF",
        "Roles": Array [
          Object {
            "Ref": "ASGInstanceRoleE263A41B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ASGInstanceRoleE263A41B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "test/ASG",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ASGInstanceSecurityGroup0525485D": Object {
      "Properties": Object {
        "GroupDescription": "test/ASG/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "test/ASG",
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "ASGInstanceSecurityGroupfromtestBastionHostInstanceSecurityGroupC58D3CDD221270B2BC": Object {
      "Properties": Object {
        "Description": "from testBastionHostInstanceSecurityGroupC58D3CDD:22",
        "FromPort": 22,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "ASGInstanceSecurityGroup0525485D",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "BastionHostInstanceSecurityGroupE75D4274",
            "GroupId",
          ],
        },
        "ToPort": 22,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "ASGInstanceSecurityGroupfromtestLBSecurityGroup6724297180E78A9185": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "FromPort": 80,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "ASGInstanceSecurityGroup0525485D",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "LBSecurityGroup8A41EA2B",
            "GroupId",
          ],
        },
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "ASGLaunchConfigC00AF12B": Object {
      "DependsOn": Array [
        "ASGInstanceRoleDefaultPolicy7636D8BF",
        "ASGInstanceRoleE263A41B",
      ],
      "Properties": Object {
        "IamInstanceProfile": Object {
          "Ref": "ASGInstanceProfile0A2834D7",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.nano",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "ASGInstanceSecurityGroup0525485D",
              "GroupId",
            ],
          },
        ],
        "UserData": Object {
          "Fn::Base64": Object {
            "Fn::Join": Array [
              "",
              Array [
                "#!/bin/bash
mkdir -p $(dirname '/tmp/",
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                "')
aws s3 cp 's3://",
                Object {
                  "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3Bucket478064AC",
                },
                "/",
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                "' '/tmp/",
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                "'
set -e
chmod +x '/tmp/",
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                "'
'/tmp/",
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb555013a04181fcdacd70c6f4b75442f710730fc69403dadc48a227bda770ceeS3VersionKey914596F9",
                        },
                      ],
                    },
                  ],
                },
                "' --verbose -y",
              ],
            ],
          },
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "ASGScalingPolicyCpuUtilization970257E9": Object {
      "Properties": Object {
        "AutoScalingGroupName": Object {
          "Ref": "ASG46ED3070",
        },
        "PolicyType": "TargetTrackingScaling",
        "TargetTrackingConfiguration": Object {
          "PredefinedMetricSpecification": Object {
            "PredefinedMetricType": "ASGAverageCPUUtilization",
          },
          "TargetValue": 50,
        },
      },
      "Type": "AWS::AutoScaling::ScalingPolicy",
    },
    "ASGScalingPolicyIncomingBytes22B7218D": Object {
      "Properties": Object {
        "AutoScalingGroupName": Object {
          "Ref": "ASG46ED3070",
        },
        "PolicyType": "TargetTrackingScaling",
        "TargetTrackingConfiguration": Object {
          "PredefinedMetricSpecification": Object {
            "PredefinedMetricType": "ASGAverageNetworkIn",
          },
          "TargetValue": 10485760,
        },
      },
      "Type": "AWS::AutoScaling::ScalingPolicy",
    },
    "ASGScalingPolicyOutgoingBytes095170F2": Object {
      "Properties": Object {
        "AutoScalingGroupName": Object {
          "Ref": "ASG46ED3070",
        },
        "PolicyType": "TargetTrackingScaling",
        "TargetTrackingConfiguration": Object {
          "PredefinedMetricSpecification": Object {
            "PredefinedMetricType": "ASGAverageNetworkOut",
          },
          "TargetValue": 10485760,
        },
      },
      "Type": "AWS::AutoScaling::ScalingPolicy",
    },
    "ASGScalingPolicyRPS9D896B6C": Object {
      "DependsOn": Array [
        "LBListener49E825B4",
        "LBListenerTargetsGroup76EF81E8",
      ],
      "Properties": Object {
        "AutoScalingGroupName": Object {
          "Ref": "ASG46ED3070",
        },
        "PolicyType": "TargetTrackingScaling",
        "TargetTrackingConfiguration": Object {
          "PredefinedMetricSpecification": Object {
            "PredefinedMetricType": "ALBRequestCountPerTarget",
            "ResourceLabel": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Fn::Select": Array [
                      1,
                      Object {
                        "Fn::Split": Array [
                          "/",
                          Object {
                            "Ref": "LBListener49E825B4",
                          },
                        ],
                      },
                    ],
                  },
                  "/",
                  Object {
                    "Fn::Select": Array [
                      2,
                      Object {
                        "Fn::Split": Array [
                          "/",
                          Object {
                            "Ref": "LBListener49E825B4",
                          },
                        ],
                      },
                    ],
                  },
                  "/",
                  Object {
                    "Fn::Select": Array [
                      3,
                      Object {
                        "Fn::Split": Array [
                          "/",
                          Object {
                            "Ref": "LBListener49E825B4",
                          },
                        ],
                      },
                    ],
                  },
                  "/",
                  Object {
                    "Fn::GetAtt": Array [
                      "LBListenerTargetsGroup76EF81E8",
                      "TargetGroupFullName",
                    ],
                  },
                ],
              ],
            },
          },
          "TargetValue": 60000,
        },
      },
      "Type": "AWS::AutoScaling::ScalingPolicy",
    },
    "BastionHost30F9ED05": Object {
      "DependsOn": Array [
        "BastionHostInstanceRoleDefaultPolicy17347525",
        "BastionHostInstanceRoleDD3FA5F1",
      ],
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "IamInstanceProfile": Object {
          "Ref": "BastionHostInstanceProfile770FCA07",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.nano",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "BastionHostInstanceSecurityGroupE75D4274",
              "GroupId",
            ],
          },
        ],
        "SubnetId": "s-12345",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "BastionHost",
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "BastionHostInstanceProfile770FCA07": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "BastionHostInstanceRoleDD3FA5F1",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "BastionHostInstanceRoleDD3FA5F1": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "BastionHost",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BastionHostInstanceRoleDefaultPolicy17347525": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ssmmessages:*",
                "ssm:UpdateInstanceInformation",
                "ec2messages:*",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BastionHostInstanceRoleDefaultPolicy17347525",
        "Roles": Array [
          Object {
            "Ref": "BastionHostInstanceRoleDD3FA5F1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BastionHostInstanceSecurityGroupE75D4274": Object {
      "Properties": Object {
        "GroupDescription": "test/BastionHost/Resource/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "114.114.192.168/32",
            "Description": "SSH access",
            "FromPort": 22,
            "IpProtocol": "tcp",
            "ToPort": 22,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "BastionHost",
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "LB8A12904C": Object {
      "Properties": Object {
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "LBSecurityGroup8A41EA2B",
              "GroupId",
            ],
          },
        ],
        "Subnets": Array [
          "s-12345",
          "s-67890",
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "LBListener49E825B4": Object {
      "Properties": Object {
        "DefaultActions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "LBListenerTargetsGroup76EF81E8",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": Object {
          "Ref": "LB8A12904C",
        },
        "Port": 80,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "LBListenerTargetsGroup76EF81E8": Object {
      "Properties": Object {
        "Port": 80,
        "Protocol": "HTTP",
        "TargetType": "instance",
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "LBSecurityGroup8A41EA2B": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB testLB7DC831DC",
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "LBSecurityGrouptotestASGInstanceSecurityGroup7EEC878A80B6A50771": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "ASGInstanceSecurityGroup0525485D",
            "GroupId",
          ],
        },
        "FromPort": 80,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "LBSecurityGroup8A41EA2B",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
  },
}
`;
